# Kunskapskontroll 1
## Databastyper

**OBS!** Alla uppgifter som inte står under rubriken **Väl godkänt** krävs för **Godkänt** på kunskapskontrollen.


1. Förklara övergripande vad NoSQL är.


Svar: NoSQL är en typ av databashanteringssytem som är lite olikt från en SQLdatabas. NoSQL brukar var bättre lämpade för stora mängder ostrukturerad eller halvstruktruerad data. NoSQL databaser används oftast i moderna applikationer när man arbetar med hög prestanda, skalbarhet och hantering av stora mängder data. Typ big data, molntjänster eller realtidanlyser av data. 

2. Vilken typ av data kan en NoSQL-databas hantera som inte en
relationell databas kan?


Svar: NoSQL databaser kräver inte en fördefinerad struktur över tabeller, kolumer och relationer. Med NoSQL är det där med möjligt att lagra både ostruktuerad data och semistrukturerad data. Till exemple JSON och BSON-dokument. Dokumentdatabaser som MongoDB kan lagra och hantera JSON och BSON format direkt, vilket gör dem mer flexibla än i en relationsdatabas som kräver en fast tabellstruktur.


3. Beskriv kortfattat vad en nyckel/värde-databas är.


Svar: Key-value eller nyckel/värde databas är data kan lagras i nycklar och värden. I en nyckel/värde databas är det endast möjligt att att utföra queries via nycklar och därför behöver dem vara unika. Strukturen i nyckel/värde databas möjliggör lagring av data utan en fördefinerad struktur. Detta medför att databasen kan skalas upp utan att sktrukturen behövs förändras. En fördel med nyckel/värde databaser är när man ska använda sig av snabba queries över enbart nycklar är fyster. Till exempel i en produkt-katalog där man snabbt ska kunna ta ut en produkt. 

4. Beskriv kortfattat vad en dokumentdatabas är.


Svar: Dokumentdatabaser är en typ av NoSQL och i grunden en nyckel/värde databas som lagrar data i ett dokumentformat, oftas en JSON, BSON, XML fil. Varje dokument innehåller en nyckel-värde-par och kan ha en flexibel struktur, vilket gör det möjligt att lagra olika typer va data utan att behöva följa en viss struktur. Typ av program man kan använda till en dokumentdatabas är mongoDB, som är ett verktyg som går att installalera i python. 

5. Beskriv kortfattat vad en bredkolumndatabas är.


Svar: Bredkolumsdatabas är databaser som lagrar data i rader och inte i kolumner. En skillnad är att varje rad kan innehålla olika kolumner. Rader med samma koluner grupperas i kolumner-familjer. Kolumn-familjer lagras oftas på olika diskar. Varje rad behöver inte ha samma kolumner, vilket gör databasen mer anpassningsbar. Apache Cassandra är ett populärt pogram att använda inom bredkolumnsdatabaser. 


6. Beskriv kortfattat vad en grafdatabas är.

Svar: Grafdatabas är en typ av NoSQL-databas som är optimerad för att lagra och hantera relationer mellan data genom att använda noder och kanter istället för tabeller. Den är särskilt kraftfull när data har många samband, som i sociala nätverk, rekommendationssystem och bedrägeridetektion. Neo4J är en populärt program att göra grafdatabaser genom att till exempel analysera sociala nätverk. 

### Nyckel/värdedatabas

Lägg till datan i filen `kunskapskontroll_1/data/northwind/orders.csv` i en Redis-databas.

In [46]:
import pandas as pd
import redis
import json




In [47]:
PWD = open("redis.pwd", "r").read().strip()

r = redis.Redis("redis-11001.crce175.eu-north-1-1.ec2.redns.redis-cloud.com", port= 11001, password=PWD, decode_responses=True, username="default")



In [48]:
r.set("Name", "Max-free-db")

True

In [49]:
r.get("Name")

r.delete("Name")

1

In [50]:
r.get("Name")

In [51]:
df = pd.read_csv("orders.csv", index_col=False, encoding="utf")

df.head()


OrderID            OrderDate         RequiredDate          ShippedDate  \
0    10248  1996-07-04 00:00:00  1996-08-01 00:00:00  1996-07-16 00:00:00   
1    10249  1996-07-05 00:00:00  1996-08-16 00:00:00  1996-07-10 00:00:00   
2    10250  1996-07-08 00:00:00  1996-08-05 00:00:00  1996-07-12 00:00:00   
3    10251  1996-07-08 00:00:00  1996-08-05 00:00:00  1996-07-15 00:00:00   
4    10252  1996-07-09 00:00:00  1996-08-06 00:00:00  1996-07-11 00:00:00   

                    ShipName ShipCountry ShipRegion           ShipAddress  \
0  Vins et alcools Chevalier      France        NaN    59 rue de l'Abbaye   
1         Toms Spezialitäten     Germany        NaN         Luisenstr. 48   
2              Hanari Carnes      Brazil         RJ       Rua do Paço, 67   
3       Victuailles en stock      France        NaN    2, rue du Commerce   
4           Suprêmes délices     Belgium        NaN  Boulevard Tirou, 255   

  ShipPostalCode        ShipCity  
0          51100           Reims  
1          44087         Münster  
2      05454-876  Rio de Janeiro  
3          69004            Lyon  
4         B-6000       Charleroi

In [99]:
df = df.fillna("")

In [98]:
df.index = pd.Index(pd.util.hash_pandas_object(df))

df.head()

OrderID            OrderDate         RequiredDate  \
12488491288975834216    10248  1996-07-04 00:00:00  1996-08-01 00:00:00   
6088303436691036197     10249  1996-07-05 00:00:00  1996-08-16 00:00:00   
18031524689564817092    10250  1996-07-08 00:00:00  1996-08-05 00:00:00   
5769693933804442962     10251  1996-07-08 00:00:00  1996-08-05 00:00:00   
757367270024661267      10252  1996-07-09 00:00:00  1996-08-06 00:00:00   

                              ShippedDate                   ShipName  \
12488491288975834216  1996-07-16 00:00:00  Vins et alcools Chevalier   
6088303436691036197   1996-07-10 00:00:00         Toms Spezialitäten   
18031524689564817092  1996-07-12 00:00:00              Hanari Carnes   
5769693933804442962   1996-07-15 00:00:00       Victuailles en stock   
757367270024661267    1996-07-11 00:00:00           Suprêmes délices   

                     ShipCountry ShipRegion           ShipAddress  \
12488491288975834216      France               59 rue de l'Abbaye   
6088303436691036197      Germany                    Luisenstr. 48   
18031524689564817092      Brazil         RJ       Rua do Paço, 67   
5769693933804442962       France               2, rue du Commerce   
757367270024661267       Belgium             Boulevard Tirou, 255   

                     ShipPostalCode        ShipCity  
12488491288975834216          51100           Reims  
6088303436691036197           44087         Münster  
18031524689564817092      05454-876  Rio de Janeiro  
5769693933804442962           69004            Lyon  
757367270024661267           B-6000       Charleroi

In [100]:
data = json.loads(df.to_json(orient="index", force_ascii=False))

data


{'12488491288975834216': {'OrderID': 10248,
  'OrderDate': '1996-07-04 00:00:00',
  'RequiredDate': '1996-08-01 00:00:00',
  'ShippedDate': '1996-07-16 00:00:00',
  'ShipName': 'Vins et alcools Chevalier',
  'ShipCountry': 'France',
  'ShipRegion': '',
  'ShipAddress': "59 rue de l'Abbaye",
  'ShipPostalCode': '51100',
  'ShipCity': 'Reims'},
 '6088303436691036197': {'OrderID': 10249,
  'OrderDate': '1996-07-05 00:00:00',
  'RequiredDate': '1996-08-16 00:00:00',
  'ShippedDate': '1996-07-10 00:00:00',
  'ShipName': 'Toms Spezialitäten',
  'ShipCountry': 'Germany',
  'ShipRegion': '',
  'ShipAddress': 'Luisenstr. 48',
  'ShipPostalCode': '44087',
  'ShipCity': 'Münster'},
 '18031524689564817092': {'OrderID': 10250,
  'OrderDate': '1996-07-08 00:00:00',
  'RequiredDate': '1996-08-05 00:00:00',
  'ShippedDate': '1996-07-12 00:00:00',
  'ShipName': 'Hanari Carnes',
  'ShipCountry': 'Brazil',
  'ShipRegion': 'RJ',
  'ShipAddress': 'Rua do Paço, 67',
  'ShipPostalCode': '05454-876',
  'ShipC

In [101]:
for k, v in data.items():
    r.hset(k, mapping=v)
    

In [56]:
p = r.hgetall(list(data.keys())[5])

In [57]:
pd.DataFrame(p.values(), index=p.keys())

0
OrderID                       10253
OrderDate       1996-07-10 00:00:00
RequiredDate    1996-07-24 00:00:00
ShippedDate     1996-07-16 00:00:00
ShipName              Hanari Carnes
ShipCountry                  Brazil
ShipRegion                       RJ
ShipAddress         Rua do Paço, 67
ShipPostalCode            05454-876
ShipCity             Rio de Janeiro

Ge ett exempel på hur vad man kan använda Redis-databasen du skapat till:

Svar: 

Redis är bra att använda när man behövern prestanda krävande system när man behöver hantera stora mängder data eller generellt stora database

### Dokumentdatabas

I mappen `kunskapskontroll_1/data/northwind/` finns filerna `products.csv` och `suppliers.json`.

Slå ihop `products.csv` och `suppliers.json` på ett passande sätt och spara dem som en dokumentdatabas i MongoDB.

#### Godkänt
Ta reda på:
* vilka produkter som behöver beställas (`ReorderLevel` $\gt$ `UnitsInStock` $ + $ `UnitsOnOrder`)
* vem man ska kontakta för att beställa fler, och
* hur man kontaktar personen


In [216]:
import json
import pandas as pd
from pymongo.mongo_client import MongoClient
from pymongo.server_api import ServerApi
import streamlit as st
import csv

In [128]:
PWD = open("mongodb.pwd", "r").read().strip()

uri = f"mongodb+srv://Kunskapskontoll_1:maxdb@cluster0.vzhgs.mongodb.net/?retryWrites=true&w=majority&appName=Cluster0"

client = MongoClient(uri, server_api=ServerApi('1'))

try:
    client.admin.command('ping')
    print("pinged your deploment. You successfully connected to MongoDB!")
except Exception as e:
    print(e)


pinged your deploment. You successfully connected to MongoDB!


In [129]:
database = client["Northwind"]
collection = database["Products"]

In [130]:
products = pd.read_csv("products.csv", index_col=False)
suppliers = pd.read_json("suppliers.json")

In [131]:
products.head()

ProductID  SupplierID                   ProductName  UnitPrice  \
0          1           1                          Chai       18.0   
1          2           1                         Chang       19.0   
2          3           1                 Aniseed Syrup       10.0   
3          4           2  Chef Anton's Cajun Seasoning       22.0   
4          6           3  Grandma's Boysenberry Spread       25.0   

   UnitsInStock  UnitsOnOrder  ReorderLevel  
0            39             0            10  
1            17            40            25  
2            13            70            25  
3            53             0             0  
4           120             0            25

In [164]:
products.drop("SupplierID", axis = 1)

ProductID                      ProductName  UnitPrice  UnitsInStock  \
0           1                             Chai      18.00            39   
1           2                            Chang      19.00            17   
2           3                    Aniseed Syrup      10.00            13   
3           4     Chef Anton's Cajun Seasoning      22.00            53   
4           6     Grandma's Boysenberry Spread      25.00           120   
..        ...                              ...        ...           ...   
64         73                       Röd Kaviar      15.00           101   
65         74                    Longlife Tofu      10.00             4   
66         75             Rhönbräu Klosterbier       7.75           125   
67         76                     Lakkalikööri      18.00            57   
68         77  Original Frankfurter grüne Soße      13.00            32   

    UnitsOnOrder  ReorderLevel  \
0              0            10   
1             40            25   
2             70            25   
3              0             0   
4              0            25   
..           ...           ...   
64             0             5   
65            20             5   
66             0            25   
67             0            20   
68             0            15   

                                      SupplierDetails  
0   {'SupplierID': 1, 'CompanyName': 'Exotic Liqui...  
1   {'SupplierID': 1, 'CompanyName': 'Exotic Liqui...  
2   {'SupplierID': 1, 'CompanyName': 'Exotic Liqui...  
3   {'SupplierID': 2, 'CompanyName': 'New Orleans ...  
4   {'SupplierID': 3, 'CompanyName': 'Grandma Kell...  
..                                                ...  
64  {'SupplierID': 17, 'CompanyName': 'Svensk Sjöf...  
65  {'SupplierID': 4, 'CompanyName': 'Tokyo Trader...  
66  {'SupplierID': 12, 'CompanyName': 'Plutzer Leb...  
67  {'SupplierID': 23, 'CompanyName': 'Karkki Oy',...  
68  {'SupplierID': 12, 'CompanyName': 'Plutzer Leb...  

[69 rows x 7 columns]

In [132]:
suppliers.head()

SupplierID                         CompanyName                 ContactName  \
0           1                      Exotic Liquids            Charlotte Cooper   
1           2          New Orleans Cajun Delights               Shelley Burke   
2           3           Grandma Kelly's Homestead               Regina Murphy   
3           4                       Tokyo Traders                Yoshi Nagase   
4           5  Cooperativa de Quesos 'Las Cabras'  Antonio del Valle Saavedra   

            Phone  
0  (171) 555-2222  
1  (100) 555-4822  
2  (313) 555-5735  
3  (03) 3555-5011  
4  (98) 598 76 54

In [133]:
products["SupplierDetails"] = \
products.apply(
    lambda x: json.loads(
        suppliers.query(f"SupplierID == {x.SupplierID}")
        .to_json(orient="records"))[0],
        axis = 1)


In [134]:
products["SupplierDetails"]

0     {'SupplierID': 1, 'CompanyName': 'Exotic Liqui...
1     {'SupplierID': 1, 'CompanyName': 'Exotic Liqui...
2     {'SupplierID': 1, 'CompanyName': 'Exotic Liqui...
3     {'SupplierID': 2, 'CompanyName': 'New Orleans ...
4     {'SupplierID': 3, 'CompanyName': 'Grandma Kell...
                            ...                        
64    {'SupplierID': 17, 'CompanyName': 'Svensk Sjöf...
65    {'SupplierID': 4, 'CompanyName': 'Tokyo Trader...
66    {'SupplierID': 12, 'CompanyName': 'Plutzer Leb...
67    {'SupplierID': 23, 'CompanyName': 'Karkki Oy',...
68    {'SupplierID': 12, 'CompanyName': 'Plutzer Leb...
Name: SupplierDetails, Length: 69, dtype: object

In [166]:
products.drop("SupplierID", axis = 1, inplace=True)

In [167]:
products_data = json.loads(products.to_json(orient="records"))

In [168]:
products_data

[{'ProductID': 1,
  'ProductName': 'Chai',
  'UnitPrice': 18.0,
  'UnitsInStock': 39,
  'UnitsOnOrder': 0,
  'ReorderLevel': 10,
  'SupplierDetails': {'SupplierID': 1,
   'CompanyName': 'Exotic Liquids',
   'ContactName': 'Charlotte Cooper',
   'Phone': '(171) 555-2222'}},
 {'ProductID': 2,
  'ProductName': 'Chang',
  'UnitPrice': 19.0,
  'UnitsInStock': 17,
  'UnitsOnOrder': 40,
  'ReorderLevel': 25,
  'SupplierDetails': {'SupplierID': 1,
   'CompanyName': 'Exotic Liquids',
   'ContactName': 'Charlotte Cooper',
   'Phone': '(171) 555-2222'}},
 {'ProductID': 3,
  'ProductName': 'Aniseed Syrup',
  'UnitPrice': 10.0,
  'UnitsInStock': 13,
  'UnitsOnOrder': 70,
  'ReorderLevel': 25,
  'SupplierDetails': {'SupplierID': 1,
   'CompanyName': 'Exotic Liquids',
   'ContactName': 'Charlotte Cooper',
   'Phone': '(171) 555-2222'}},
 {'ProductID': 4,
  'ProductName': "Chef Anton's Cajun Seasoning",
  'UnitPrice': 22.0,
  'UnitsInStock': 53,
  'UnitsOnOrder': 0,
  'ReorderLevel': 0,
  'SupplierDet

In [287]:
collection.delete_many({})

DeleteResult({'n': 69, 'electionId': ObjectId('7fffffff000000000000008c'), 'opTime': {'ts': Timestamp(1739526815, 12), 't': 140}, 'ok': 1.0, '$clusterTime': {'clusterTime': Timestamp(1739526815, 12), 'signature': {'hash': b'\x9c\x01\xf2\xbc\xdfo\xdfDF5\x96=\xea+\x8d\xe3\x87\x1a\xbc\x90', 'keyId': 7413779418855768081}}, 'operationTime': Timestamp(1739526815, 12)}, acknowledged=True)

In [288]:
collection.insert_many(products_data)

InsertManyResult([ObjectId('67aa04dbfc8b0d3c92e159d9'), ObjectId('67aa04dbfc8b0d3c92e159da'), ObjectId('67aa04dbfc8b0d3c92e159db'), ObjectId('67aa04dbfc8b0d3c92e159dc'), ObjectId('67aa04dbfc8b0d3c92e159dd'), ObjectId('67aa04dbfc8b0d3c92e159de'), ObjectId('67aa04dbfc8b0d3c92e159df'), ObjectId('67aa04dbfc8b0d3c92e159e0'), ObjectId('67aa04dbfc8b0d3c92e159e1'), ObjectId('67aa04dbfc8b0d3c92e159e2'), ObjectId('67aa04dbfc8b0d3c92e159e3'), ObjectId('67aa04dbfc8b0d3c92e159e4'), ObjectId('67aa04dbfc8b0d3c92e159e5'), ObjectId('67aa04dbfc8b0d3c92e159e6'), ObjectId('67aa04dbfc8b0d3c92e159e7'), ObjectId('67aa04dbfc8b0d3c92e159e8'), ObjectId('67aa04dbfc8b0d3c92e159e9'), ObjectId('67aa04dbfc8b0d3c92e159ea'), ObjectId('67aa04dbfc8b0d3c92e159eb'), ObjectId('67aa04dbfc8b0d3c92e159ec'), ObjectId('67aa04dbfc8b0d3c92e159ed'), ObjectId('67aa04dbfc8b0d3c92e159ee'), ObjectId('67aa04dbfc8b0d3c92e159ef'), ObjectId('67aa04dbfc8b0d3c92e159f0'), ObjectId('67aa04dbfc8b0d3c92e159f1'), ObjectId('67aa04dbfc8b0d3c92e159

In [297]:
# Query för att få ut vilka produktersom behövs beställa hem
query = {
    "$expr": {
        "$gt": [
            "$ReorderLevel",
            {"$sum": ["$UnitsInStock", "$UnitsOnOrder"] }
        ]
    }
}

products_to_order = collection.find(query)

for product in products_to_order:
    print(product)

ServerSelectionTimeoutError: SSL handshake failed: cluster0-shard-00-01.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms),SSL handshake failed: cluster0-shard-00-00.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms),SSL handshake failed: cluster0-shard-00-02.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms), Timeout: 30s, Topology Description: <TopologyDescription id: 67a9eaa2fc8b0d3c92e15993, topology_type: ReplicaSetNoPrimary, servers: [<ServerDescription ('cluster0-shard-00-00.vzhgs.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: cluster0-shard-00-00.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>, <ServerDescription ('cluster0-shard-00-01.vzhgs.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: cluster0-shard-00-01.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>, <ServerDescription ('cluster0-shard-00-02.vzhgs.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: cluster0-shard-00-02.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>

In [286]:
# Att man ser 
pipeline = [
    {
        "$project": {
            "_id": 1,
            "CompanyName": "$SupplierDetail.CompanyName",
            "Phone": "$SupplierDetail.Phone",
            "ContactName": "$SupplierDetail.ContactName"
        }
    }
]

suppliers = list(collection.aggregate(pipeline))
if suppliers:
    collection.insert_many(suppliers)

    print("SupplierDetails har extraherats och sparats i 'suppliers' samlingen.")

SupplierDetails har extraherats och sparats i 'suppliers' samlingen.


In [294]:
pipeline = [
    {
        "$lookup": {
            "from": "SuppliersDetails",  
            "localField": "SupplierDetails.supplierID",  
            "foreignField": "supplierID",  
            "as": "supplier_Details"
        }
    },
    {
        "$addFields": {
            "supplierdetail": { "$arrayElemAt": ["$supplier_Details", 0] } 
        }
    },
    {
        "$project": { "supplier_info": 0 } 
    }
]

orders_with_supplier = list(collection.aggregate(pipeline))


ServerSelectionTimeoutError: SSL handshake failed: cluster0-shard-00-01.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms),SSL handshake failed: cluster0-shard-00-00.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms),SSL handshake failed: cluster0-shard-00-02.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms), Timeout: 30s, Topology Description: <TopologyDescription id: 67a9eaa2fc8b0d3c92e15993, topology_type: ReplicaSetNoPrimary, servers: [<ServerDescription ('cluster0-shard-00-00.vzhgs.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: cluster0-shard-00-00.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>, <ServerDescription ('cluster0-shard-00-01.vzhgs.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: cluster0-shard-00-01.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>, <ServerDescription ('cluster0-shard-00-02.vzhgs.mongodb.net', 27017) server_type: Unknown, rtt: None, error=AutoReconnect('SSL handshake failed: cluster0-shard-00-02.vzhgs.mongodb.net:27017: [SSL: TLSV1_ALERT_INTERNAL_ERROR] tlsv1 alert internal error (_ssl.c:1018) (configured timeouts: socketTimeoutMS: 20000.0ms, connectTimeoutMS: 20000.0ms)')>]>


#### Väl godkänt
Skapa en Streamlit-app som visar produkter som behöver beställas och talar om vem man ska ringa för att beställa fler. Appen kan vara en enkel lista över produkter, eller innehålla fler funktioner (QR-kod-scanner?).

Kolla Lagerstatus.py filen i Github

### Grafdatabas
**OBS** Fråga 1-3 nedan är desamma för både Godkänt och Väl godkänt. Det som skiljer sig är hur du väljer att skapa databasen samt fråga 4.

#### Godkänt
* Läs in `csv`-filerna i `kunskapskontroll_1/data/movies/prepared/` i Neo4j och skapa en datamodell.
* Svara på frågorna och utför instruktionerna nedan. Klistra in dina svar som JSON eller bilder, beroende på om svaret är i data- eller graf-form.
    1. Vilka är de 10 vanligaste kombinationerna av genrer?
    2. Vilka är de fem vanligaste genrerna på filmer regisserade av Steven Spielberg, och hur många filmer är i de genrerna?
    3. Det verkar som om de har missat att Viggo Mortensen är med i Sagan om Ringen-filmerna. Skapa `[:ACTED_IN]`-relationer mellan Viggo och alla tre *The Lord of the Rings*-filmerna. Visa en graf med Viggo i mitten och relationer till de tre filmerna.

#### Väl godkänt
* Bearbeta datan i `kunskapskontroll_1/data/movies/movies.json` så att den passar till att läsas in i Neo4j. Processen ska gå att följa i kodceller i notebooken. Du får gärna titta på filerna i `kunskapskontroll_1/data/movies/prepared/` för att få en idé om hur resultatet ska se ut.
* Läs in data in Neo4j.
* Svara på frågorna och utför instruktionerna nedan. Klistra in dina svar som JSON eller bilder, beroende på om svaret är i data- eller graf-form.
    1. Vilka är de 10 vanligaste kombinationerna av genrer?
    2. Vilka är de fem vanligaste genrerna på filmer regisserade av Steven Spielberg, och hur många filmer är i de genrerna?
    3. Det verkar som om de har missat att Viggo Mortensen är med i Sagan om Ringen-filmerna. Skapa `[:ACTED_IN]`-relationer mellan Viggo och alla tre *The Lord of the Rings*-filmerna. Visa en graf med Viggo i mitten och relationer till de tre filmerna.
    4. Hitta på minst en egen fråga och svara på den.



In [243]:
from neo4j import GraphDatabase

with open("Neo4j.pwd") as f:
    pwd_neo4j = f.read().strip()

# URI funkade inte så fick använda mig av +SSC 
URI = "neo4j+ssc://199972c7.databases.neo4j.io"
AUTH = ("neo4j", pwd_neo4j)

with GraphDatabase.driver(URI, auth=AUTH) as driver:
    driver.verify_connectivity()


In [244]:
with open("movies.json", "r") as f:
    movies = json.load(f)

In [274]:



with open("movies.json", "r", encoding="utf-8") as file:
    movies_data = json.load(file)

# Skapa unika ID:n för actors, directors och genres
actor_dict = {}
director_dict = {}
genre_dict = {}

# Filinnehåll
movies_list = []
actors_movies_list = []
actors_list = []
directors_list = []
genres_movies_list = []
genres_list = []

# Generera data för CSV-filer
for movie in movies_data:
    movie_id = movie["movie_id"]
    
    
    movies_list.append([
        movie_id, movie["movie_title"].strip(), movie["title_year"], 
        movie["director_name"], movie["duration"], movie["gross"], 
        movie["num_voted_users"], movie["language"], movie["country"], 
        movie["content_rating"], movie["budget"], movie["title_year"], 
        movie["imdb_score"], movie["aspect_ratio"]
    ])
    
    
    if movie["director_name"] not in director_dict.keys():
        director_id = len(director_dict) + 1
        director_dict[movie["director_name"]] = director_id
        directors_list.append([director_id, movie["director_name"]])

    # Actors.csv & Actors_movies.csv
    for actor_name in movie["actors"]:
        if actor_name not in actor_dict.keys():
            actor_id = len(actor_dict) + 1
            actor_dict[actor_name] = actor_id
            actors_list.append([actor_id, actor_name])
        actors_movies_list.append([actor_dict[actor_name], movie_id])

    # Genres.csv & Genres_movies.csv
    for genre in movie["genres"]:
        if genre not in genre_dict:
            genre_id = len(genre_dict) + 1
            genre_dict[genre] = genre_id
            genres_list.append([genre_id, genre])
        genres_movies_list.append([genre_dict[genre], movie_id])

# Funktion för att skriva CSV-filer
def write_csv(filename, data, headers):
    with open(filename, "w", newline="", encoding="utf-8") as file:
        writer = csv.writer(file)
        writer.writerow(headers)
        writer.writerows(data)

# Skapa CSV-filer
write_csv("movies.csv", movies_list, ["movie_id", "movie_title", "year", "director", "duration", "gross", "num_voted_users", "language", "country", "content_rating", "budget", "title_year", "imdb_score", "aspect_ratio"])
write_csv("actors_movies.csv", actors_movies_list, ["actor_id", "movie_id"])
write_csv("actors.csv", actors_list, ["actor_id", "actor_name"])
write_csv("directors.csv", directors_list, ["director_id", "director_name"])
write_csv("genres_movies.csv", genres_movies_list, ["genre_id", "movie_id"])
write_csv("genres.csv", genres_list, ["genre_id", "genre"])


#  Vilka är de 10 vanligaste kombinationerna av genrer?
MATCH (m:Movies)-[:HAS_GENRE]->(g:Genres) with m, collect(g.genre) as genres RETURN genres, count(*) AS count ORDER BY count DESC LIMIT 10;

[
  {
    "genres": [
      "Drama"
    ],
    "count": 233
  },
  {
    "genres": [
      "Comedy"
    ],
    "count": 205
  },
  {
    "genres": [
      "Comedy",
      "Drama"
    ],
    "count": 189
  },
  {
    "genres": [
      "Comedy",
      "Romance",
      "Drama"
    ],
    "count": 185
  },
  {
    "genres": [
      "Comedy",
      "Romance"
    ],
    "count": 157
  },
  {
    "genres": [
      "Romance",
      "Drama"
    ],
    "count": 150
  },
  {
    "genres": [
      "Thriller",
      "Drama",
      "Crime"
    ],
    "count": 98
  },
  {
    "genres": [
      "Horror"
    ],
    "count": 67
  },
  {
    "genres": [
      "Action",
      "Thriller",
      "Drama",
      "Crime"
    ],
    "count": 65
  },
  {
    "genres": [
      "Drama",
      "Crime"
    ],
    "count": 63
  }
]

# Vilka är de fem vanligaste genrerna på filmer regisserade av Steven Spielberg, och hur många filmer är i de genrerna?

MATCH (d:DIRECTORS {director_name: "Steven Spielberg"})-[:DIRECTED]->(m:Movies)MATCH (m)-[:HAS_GENRE]->(g:Genres) RETURN g.genre as Genre, COUNT(m) as Count ORDER BY Count DESC LIMIT 5;

[
  {
    "Genre": "Drama",
    "Count": 13
  },
  {
    "Genre": "Adventure",
    "Count": 12
  },
  {
    "Genre": "Action",
    "Count": 9
  },
  {
    "Genre": "Sci-Fi",
    "Count": 7
  },
  {
    "Genre": "Thriller",
    "Count": 6
  }
]

#3. Det verkar som om de har missat att Viggo Mortensen är med i Sagan om Ringen-filmerna. Skapa `[:ACTED_IN]`-relationer mellan Viggo och alla tre *The Lord of the Rings*-filmerna. Visa en graf med Viggo i mitten och relationer till de tre filmerna.


MATCH (a: Actors {actor_name: "Viggo Mortensen"}) MATCH (m1: Movies {movie_title: "The Lord of the Rings: The Fellowship of the Ring"}) MATCH (m2: Movies {movie_title: "The Lord of the Rings: The Two Towers"}) MATCH (m3: Movies {movie_title: "The Lord of the Rings: The Return of the King"}) MERGE (a)-[:ACTED_IN]->(m1) MERGE (a)-[:ACTED_IN]->(m2) MERGE (a)-[:ACTED_IN]->(m3)





# Enklare Cypher query för att lägga till "Viggo" i databasen.

MATCH path=(a: Actors {actor_name: "Viggo Mortensen"}) -[:ACTED_IN]->(m:Movies) WHERE m.movie_title CONTAINS "Lord of the" RETURN path

# 4. Hitta på minst en egen fråga och svara på den.

Vilka "directors" som har flest filmer med högst imdb_score

MATCH (d:DIRECTORS)-[:DIRECTED]->(m:Movies)
WITH d, COUNT(m) AS num_movies, SUM(m.imdb_score) AS total_score
ORDER BY num_movies DESC, total_score DESC
RETURN d.director_name, num_movies, total_score
LIMIT 10;


  {
    "d.director_name": "M. Night Shyamalan",
    "num_movies": 169,
    "total_score": 1015.8000000000001
  },
  {
    "d.director_name": "Richard Donner",
    "num_movies": 149,
    "total_score": 915.4999999999999
  },
  {
    "d.director_name": "Edward Zwick",
    "num_movies": 142,
    "total_score": 889.6000000000001
  },
  {
    "d.director_name": "Ang Lee",
    "num_movies": 140,
    "total_score": 884.2000000000003
  },
  {
    "d.director_name": "Dean Parisot",
    "num_movies": 135,
    "total_score": 817.9999999999993
  },
  {
    "d.director_name": "Rob Minkoff",
    "num_movies": 130,
    "total_score": 752.1999999999996
  },
  {
    "d.director_name": "Alex Proyas",
    "num_movies": 128,
    "total_score": 787.9000000000001
  },
  {
    "d.director_name": "Lee Tamahori",
    "num_movies": 127,
    "total_score": 761.0
  },
  

## Vad ska lämnas in på Learnpoint?

Deadline: 14/2 kl. 17.00

En länk till ett Github-repository som innehåller:
* en kopia av den här notebooken med dina svar på uppgifterna. **Använd kod-celler för kod och markdown-celler för text!**
* bilder på grafer så att de kan visas i notebooken
* VG: om du gjort en Streamlit-app i dokumentdatabas-delen ska där också finnas en `app.py` eller motsvarande med koden för appen.

That's it!